### PR TITLE
Reland: Make GetDefaultEventEngine return a shared_ptr

### PR DESCRIPTION
**Pull request by drfloob**

### CODE REVIEW

File: src/core/lib/event_engine/default_event_engine.cc [hunk 1]
```diff
@@ -23,22 +23,34 @@
...
+#include "src/core/lib/gprpp/sync.h"
 
 namespace grpc_event_engine {
 namespace experimental {
 
 namespace {
 std::atomic<absl::AnyInvocable<std::unique_ptr<EventEngine>()>*>
     g_event_engine_factory{nullptr};
-std::atomic<EventEngine*> g_event_engine{nullptr};
+grpc_core::NoDestructSingleton<grpc_core::Mutex> g_mu;
```
Comment (ctiller) [inline src/core/lib/event_engine/default_event_engine.cc:38]: This should be `NoDestruct<grpc_core::Mutex> g_mu;` - same for the next line.
I'm gonna send out a PR in a moment to fix up the API problem.

**Comment by ctiller** [comment]
lgtm

File: src/core/lib/event_engine/default_event_engine.cc (hunk 1 above)
Comment (drfloob) [reply src/core/lib/event_engine/default_event_engine.cc]: Thanks, fixed.

**Comment by drfloob** [comment]
Thanks! Please click approve when you get a chance, I'm going to run a broad set of tests on it and merge tomorrow if all looks good.

**Review by ctiller — APPROVED** [review]

**Comment by drfloob** [comment]
The TGP was clean after deflaking once. I'll merge this first thing tomorrow.

**Pull request merged**